### PR TITLE
Add support for browserify.

### PR DESCRIPTION
**Pull request by vicapow**
Here's my first go at fixing the browserify issues. I haven't tested in IE10/11 but because of lack of blob URL support, this will probably still break.

see: https://github.com/mapbox/mapbox-gl-js/issues/787 for context

### CODE REVIEW

**Comment by jfirebaugh** [comment]
Thanks @vicapow, this looks like a good start. [Here's a patch that fixes lint and test issues](https://gist.github.com/jfirebaugh/dd305b80dd78afbd5e4e).

We do need to maintain IE compatibility however -- is there a fallback strategy for that?

**Comment by vicapow** [comment]
Nice, thanks @jfirebaugh. 

One possible workaround for IE10/11 is to have a `webworker-helper.js` file hosted somewhere (maybe a mapbox server?) that only contains something like the following:

```js
// webworker-helper.js
self.onmessage = function(e) {
    eval(e.browserifyScripts);
};
```
see: http://stackoverflow.com/questions/10343913/how-to-create-a-web-worker-from-a-string

Presumably, this could also be configured to use a local server instead.

If IE Edge fixes this issue, it'll start doing what all the other browsers do.

**Comment by vicapow** [comment]
On top of that, there could be a separate bootstrap file that does exactly what mapbox GL already does that's only used in the `dist/mapbox-gl.js` file that wouldn't be included in the browserify entry point.

**Comment by jfirebaugh** [comment]
I think we should go with the helper file and have `mapbox-gl.js` continue to be the same as what you get if you browserify it yourself. That means:

* Feature-detect blob URL support in `Dispatcher`.
* If not supported, fall back to helper file approach.
* Helper file lives in `dist/mapbox-gl-worker-helper.js`.
* Add helper file to `deploy.sh` and also document the deployment requirements for non-Mapbox CDN usage.

@vicapow Want to give this a shot?

**Comment by vicapow** [comment]
@jfirebaugh I'd be happy to :grinning:

**Comment by vicapow** [comment]
So it turns out IE11 _does_ handle blob urls passed to workers correctly! It's only IE10 that doesn't. And IE10 doesn't support WebGL. So this PL should be good, right? :sunglasses:

**Comment by vicapow** [comment]
TL;DR I still haven't found a way to get this to work with IE.

After installing a windows VM and doing some digging myself, I'm going to back track on what I said above. Even though blob's can be passed to web workers in IE11, web workers created in this way have an undefined origin and any request for a resource on the original origin produces a ProgressEvent error because of it.

````js
var src = '\'use strict\';\n var req = new XMLHttpRequest();\n' +
'req.onload = function requestListener() {\n' +
  'console.log(this.status);\n' +
  'console.log(this.statusCode);\n' +
  'console.log(this.responseText);\n' +
'};\n' +
'req.onerror = function errorListener(e) {\n' +
  'console.log(\'an error occured\');\n' +
  'console.log(e.currentTarget.status);\n' +
  'console.log(e.currentTarget.statusCode);\n' +
  'console.log(e);\n' +
'};\n' +
// `onerror` callback is called with a "ProgressEvent" and status code 0.
'req.open(\'get\', \'http://localhost:1338/data.txt\', true);\n' +
'req.send();\n' +
'console.log(\'origin\', self.location.origin);\n'; // undefined in IE

var worker = new Worker(URL.createObjectURL(
    new Blob([src], {type: 'text/javascript'})
));
````

Since one of the main points for having webworkers in mapbox-GL is to offload requesting and parsing of large geojson files, I suspect this is a blocker. :cry: 

On top of that, the "worker-helper.js" approach doesn't work because IE will through a `SecurityError` when attempting to create a webworker from a different origin than the hosted page. This means we could have a worker-helper.js file hosted in mapbox servers being run from sites not on the same domain 
:sob: 

````js
// SecurityError
var worker = new window.Worker('http://someotherdomain:1338/worker-helper.js');
````

**Comment by jfirebaugh** [comment]
> On top of that, the "worker-helper.js" approach doesn't work because IE will through a SecurityError when attempting to create a webworker from a different origin than the hosted page.

This should be possible with `createObjectURL` + `new Blob` + `importScripts`. Check out the current approach for initializing the worker, which works on a third-party site in IE with `mapbox-gl.js` hosted on a `mapbox.com` domain.

**Comment by vicapow** [comment]
Not sure why CircleCl is failing but the issues I was running into earlier had to do with my local static file server :weary:. Those have all been resolved and everything works now on IE11. :laughing: 

ie11-screenshot

**Comment by vicapow** [comment]
That was a screenshot of it running standalone, here's the running browserify version: 

ie11-browserify

**Comment by jfirebaugh** [comment]
This is great, thanks @vicapow. Committed as 2adecdb1cbbd6298ee225fd2f3a0953b236120e8.

**Comment by mourner** [comment]
The bundle is now more than 2x bigger in size though :(

**Comment by jfirebaugh** [comment]
Aaaa what!? @vicapow can you check out where the size increase comes from?

**Comment by mourner** [comment]
Maybe the browserified URL module?

**Comment by vicapow** [comment]
That's no good. I'll take a look. My intuition is the same. It's probably
the URL module which likely has a bunch of sub dependencies

**Comment by vicapow** [comment]
Just for reference, the issues with minification size increases was fixed in: https://github.com/mapbox/mapbox-gl-js/pull/1238
